File: 백준/Gold/1238. 파티/파티.kt
import java.lang.Integer.min

lateinit var dist: Array<IntArray>

fun main() {
    val (N, M, X) = readLine()!!.split(" ").map { it.toInt() }
    dist = Array(N + 1) { IntArray(N + 1) { 100001 } }

    repeat(M) {
        val (s, e, c) = readLine()!!.split(" ").map { it.toInt() }
        dist[s][e] = c
    }

    floyd(N)
    val roundTrip = IntArray(N + 1) { 0 }
    for (i in 1..N) {
        roundTrip[i] = dist[i][X] + dist[X][i]
    }
    println(roundTrip.max())
}

fun floyd(n: Int) {
    for (i in 1..n) {
        dist[i][i] = 0
    }
    for (k in 1 until n + 1) {
        for (i in 1 until n + 1) {
            for (j in 1 until n + 1) {
                dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])
            }
        }
    }
}

File: 백준/Gold/1238. 파티/파티.cc
#include <stdio.h>

int main(){
	int n,m,x,i,j,k,max;
	int map[1001][1001];
	
    //입력 받기 및 초기화
	scanf("%d %d %d",&n,&m,&x);
	for(i=1;i<=n;i++){
		for(j=1;j<=n;j++)
			map[i][j] = 100000;
		map[i][i]=0;
    }
	
    //도로 길이 입력받기
	while(m--){
		scanf("%d %d %d",&i,&j,&k);
		map[i][j] = k;
	}
	
    //플로이드 와샬 알고리즘
	for(k=1;k<=n;k++){
		for(i=1;i<=n;i++)
			for(j=1;j<=n;j++){
				if(map[i][j] > map[i][k]+map[k][j])
					map[i][j] = map[i][k]+map[k][j];
			}
	}
	
    //왕복거리 최대값 찾기
	max = 0;
	for(i=1;i<=n;i++){
		if(map[i][x]+map[x][i] > max)
			max = map[i][x]+map[x][i];
	}
	printf("%d\n",max);	
	
	return 0;
}